### PR TITLE
659: Collect and test baseline profiles

### PR DESCRIPTION
**Pull request by steve-the-edwards**
For #659.

Adds in a macro-benchmark module to gather baseline profiles and benchmark impact on a scenario in the Dungeon Sample App.

Add the full profile the Dungeon sample app and add in the split profile sections to each module in the Workflow library.

Google Engineer reported 10% performance improvement testing on phone (see [here](https://github.com/tikurahul/workflow-kotlin/commit/0ebb11e91a568c63c60a9052d65da738be8bcfd5) and comments below). I will be doing my own testing on phone soon.
Testing on a Square terminal results are not yet improving with baseline profiles.

Also added:

- Updated to Gradle. 7.4
- Gradle Enterprise (with public upload) because we want to be able to diagnose gradle build issues by sharing scans.
- Commented out code for using AndroidX snapshot builds (to get the bleeding edge tools when needed).
- Updated AGP that I've left commented out, only for use when on bleeding edge.

### CODE REVIEW

File: benchmarks/dungeon-benchmark/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  defaultConfig {
+    minSdk = 23
+    targetSdk = 32
+
+    testInstrumentationRunner = "androidx.test.runner.AndroidJUnitRunner"
+  }
+
+  buildTypes {
+    create("benchmark") {
+      isDebuggable = true
```
Comment (pyricau) [inline benchmarks/dungeon-benchmark/build.gradle.kts:30]: You likely want profileable but not debuggable? Or is that not the main app, but the app that runs the test instead?

File: benchmarks/dungeon-benchmark/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      signingConfig = getByName("debug").signingConfig
+    }
+  }
+
+  targetProjectPath = ":samples:dungeon:app"
```
Comment (pyricau) [inline benchmarks/dungeon-benchmark/build.gradle.kts:35]: ah yes that's the main app, ignore my previous comment.

File: benchmarks/dungeon-benchmark/src/main/java/com/squareup/sample/dungeon/benchmark/WorkflowBaselineBenchmark.kt [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  public fun benchmarkBaselineProfiles() {
+    benchmark(CompilationMode.Partial(baselineProfileMode = BaselineProfileMode.Require))
+  }
+
+  @OptIn(ExperimentalTime::class)
+  public fun benchmark(compilationMode: CompilationMode) {
+    benchmarkRule.measureRepeated(
+      packageName = PACKAGE_NAME,
+      metrics = listOf(StartupTimingMetric(), FrameTimingMetric()),
+      iterations = 1,
```
Comment (pyricau) [inline benchmarks/dungeon-benchmark/src/main/java/com/squareup/sample/dungeon/benchmark/WorkflowBaselineBenchmark.kt:61]: a single iteration seems wrong. Should be more like 100?

File: benchmarks/dungeon-benchmark/src/main/java/com/squareup/sample/dungeon/benchmark/WorkflowBaselineBenchmark.kt [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+      startupMode = StartupMode.COLD,
+      compilationMode = compilationMode,
+      setupBlock = {
+        pressHome()
+
```
Comment (pyricau) [inline benchmarks/dungeon-benchmark/src/main/java/com/squareup/sample/dungeon/benchmark/WorkflowBaselineBenchmark.kt:66]: empty line

File: benchmarks/dungeon-benchmark/src/main/java/com/squareup/sample/dungeon/benchmark/WorkflowBaselineProfiles.kt [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+/**
+ * Shamelessly copied from the example the Google Benchmark team made on the Lottie library
+ * https://github.com/airbnb/lottie-android/pull/2005.
+ *
+ * This test actually generates the profiles.
+ */
+@RunWith(AndroidJUnit4::class)
+@OptIn(ExperimentalBaselineProfilesApi::class)
+public class WorkflowBaselineProfiles {
```
Comment (pyricau) [inline benchmarks/dungeon-benchmark/src/main/java/com/squareup/sample/dungeon/benchmark/WorkflowBaselineProfiles.kt:32]: Isn't this mostly the same code as the other test? Can we reuse some bits?

**Review by pyricau — APPROVED** [review]

File: benchmarks/dungeon-benchmark/build.gradle.kts (hunk 1 above)
Comment (steve-the-edwards) [reply benchmarks/dungeon-benchmark/build.gradle.kts:30]: This is the app that runs the test.

File: benchmarks/dungeon-benchmark/src/main/java/com/squareup/sample/dungeon/benchmark/WorkflowBaselineBenchmark.kt (hunk 3 above)
Comment (steve-the-edwards) [reply benchmarks/dungeon-benchmark/src/main/java/com/squareup/sample/dungeon/benchmark/WorkflowBaselineBenchmark.kt:61]: I was just following what the Lottie example did for this!

File: benchmarks/dungeon-benchmark/src/main/java/com/squareup/sample/dungeon/benchmark/WorkflowBaselineProfiles.kt (hunk 5 above)
Comment (steve-the-edwards) [reply benchmarks/dungeon-benchmark/src/main/java/com/squareup/sample/dungeon/benchmark/WorkflowBaselineProfiles.kt:32]: Yes we can. :)

**Comment by steve-the-edwards** [comment]
Instructions to get results off of a rooted play store device:

1. Set `agpVersion=7.3.0-alpha03` in `~/.gradle/gradle.properties`
2. `./gradlew :samples:dungeon:app:installRelease`
3. `./gradlew :benchmarks:dungeon-benchmark:connectedReleaseAndroidTest -Pandroid.testInstrumentationRunnerArguments.class="com.squareup.sample.dungeon.benchmark.WorkflowBaselineBenchmark#benchmarkNoCompilation"`
4. `./gradlew :benchmarks:dungeon-benchmark:connectedReleaseAndroidTest -Pandroid.testInstrumentationRunnerArguments.class="com.squareup.sample.dungeon.benchmark.WorkflowBaselineBenchmark#benchmarkBaselineProfiles"`

**Comment by steve-the-edwards** [comment]
So far the results are worse??

AGP 7.3-alpha03
Gradle 7.4
Latest Snapshot
WorkflowBaselineBenchmark_benchmarkNoCompilation
timeToInitialDisplayMs  min 536.4,  median 658.3,  max 748.8
frameDurationCpuMs  P50  12.1,  P90  21.1,  P95  30.9,  P99 173.7
Traces: Iteration 0 1 2 3 4 5 6 7 8 9
WorkflowBaselineBenchmark_benchmarkBaselineProfiles
timeToInitialDisplayMs  min  830.3,  median  917.6,  max 1,044.1
frameDurationCpuMs  P50   12.5,  P90   20.6,  P95   28.9,  P99  166.2
Traces: Iteration 0 1 2 3 4 5 6 7 8 9

**Comment by tikurahul** [comment]
https://github.com/tikurahul/workflow-kotlin/commit/0ebb11e91a568c63c60a9052d65da738be8bcfd5

The interesting metric is this:

```
WorkflowBaselineBenchmark_benchmarkNoCompilation
timeToInitialDisplayMs   min 227.9,   median 242.3,   max 275.3

WorkflowBaselineBenchmark_benchmarkBaselineProfiles
timeToInitialDisplayMs   min 225.2,   median 235.3,   max 246.0
```

File: benchmarks/dungeon-benchmark/build.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,52 @@
+plugins {
+  id("com.android.test")
+  id("org.jetbrains.kotlin.android")
+}
+
+// Note: We are not including our defaults from .buildscript as we not need the base Workflow
+// dependencies that those include.
+
```
Comment (RBusarow) [inline benchmarks/dungeon-benchmark/build.gradle.kts:8]: Are we also tied to these sdk versions?

**Review by RBusarow — APPROVED** [review]

**Review by tikurahul — APPROVED** [review]

File: benchmarks/dungeon-benchmark/src/main/java/com/squareup/sample/dungeon/benchmark/WorkflowBaselineProfiles.kt [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        UiScrollable(UiSelector().className("androidx.recyclerview.widget.RecyclerView"))
+      boardsList.waitForExists(UI_TIMEOUT_MS)
+      val mazeBoardSelector = UiSelector().enabled(true).clickable(true).instance(1)
+      val mazeBoard = boardsList.getChild(mazeBoardSelector)
+      mazeBoard.waitForExists(UI_TIMEOUT_MS)
+      mazeBoard.clickAndWaitForNewWindow()
+
+      val leftButttonSelector = UiSelector()
+        .resourceId("$PACKAGE_NAME:id/move_left")
+      val leftButton = device.findObject(leftButttonSelector)
```
Comment (rjrjr) [inline benchmarks/dungeon-benchmark/src/main/java/com/squareup/sample/dungeon/benchmark/WorkflowBaselineProfiles.kt:66]: usually two 't's in button.

File: gradle/libs.versions.toml [hunk 8]
```diff
@@ -1,5 +1,6 @@
 [versions]
 
+#androidTools = "7.3.0-alpha03"
```
Comment (rjrjr) [inline gradle/libs.versions.toml:3]: Snip the commented entries? Or else document them.

**Review by rjrjr — APPROVED** [review]
Congrats!

Please remember to follow up with documentation on how we update these things. Will we need to do so as part of each release? That could certainly complicate things.

**Comment by tikurahul** [comment]
One more thing I wanted to mention. In Workflows case it could be that a lot of work happens after the first frame is drawn. So time to first paint may not be the most meaningful metric. 

You might want to consider using `reportFullyDrawn()`. That way you can measure when the app was ready to interact with.

File: benchmarks/dungeon-benchmark/build.gradle.kts (hunk 6 above)
Comment (steve-the-edwards) [reply benchmarks/dungeon-benchmark/build.gradle.kts:8]: The ones specified here you mean? Or the ones in the base files in `.buildscript`? We are not tied to any for running the benchmarks per se.

**Comment by steve-the-edwards** [comment]
> Congrats!
> 
> Please remember to follow up with documentation on how we update these things. Will we need to do so as part of each release? That could certainly complicate things.

@rjrjr included `README` with instructions for how to do this and contextual explanation. Thanks for the reminder!

**Comment by tikurahul** [comment]
> Do you mean this can artificially adjust when the tests considers the app fully done and we should be calling this from > > within Library code?

Yes, that is correct. We actually emit a new metric as well.

**Pull request merged**